packages/tailwindcss/src/compat/config/resolve-config.ts

```diff
@@ -0,0 +1,141 @@
+import { segment } from '../../utils/segment'
+import { deepMerge, isPlainObject } from './deep-merge'
+import {
+  type ResolvedConfig,
+  type ResolvedThemeValue,
+  type ThemeValue,
+  type UserConfig,
+} from './types'
+
+interface ResolutionContext {
+  configs: UserConfig[]
+  theme: Record<string, ThemeValue>
+  extend: Record<string, ThemeValue[]>
+  result: ResolvedConfig
+}
+
+let minimal: ResolvedConfig = {
+  theme: {},
+}
+
+export function resolveConfig(configs: UserConfig[]): ResolvedConfig {
+  let ctx: ResolutionContext = {
+    configs,
+    theme: {},
+    extend: {},
+
+    // Start with a minimal valid, but empty config
+    result: structuredClone(minimal),
+  }
+
+  // Merge themes
+  mergeTheme(ctx)
+
+  return {
+    theme: ctx.theme as ResolvedConfig['theme'],
+  }
+}
+
+function mergeThemeExtension(
+  themeValue: ThemeValue | ThemeValue[],
+  extensionValue: ThemeValue | ThemeValue[],
+) {
+  // When we have an array of objects, we do want to merge it
+  if (Array.isArray(themeValue) && isPlainObject(themeValue[0])) {
+    return themeValue.concat(extensionValue)
+  }
+
+  // When the incoming value is an array, and the existing config is an object, prepend the existing object
+  if (
+    Array.isArray(extensionValue) &&
+    isPlainObject(extensionValue[0]) &&
+    isPlainObject(themeValue)
+  ) {
+    return [themeValue, ...extensionValue]
+  }
+
+  // Override arrays (for example for font-families, box-shadows, ...)
+  if (Array.isArray(extensionValue)) {
+    return extensionValue
+  }
+
+  // Execute default behaviour
+  return undefined
+}
+
+export interface PluginUtils {
+  theme(keypath: string, defaultValue?: any): any
+}
+
+function createThemeFn(ctx: ResolutionContext, resolveValue: (value: any) => any) {
+  return function theme(path: string): any {
+    let parts = segment(path, '.')
+    let value: any = ctx.theme
+
+    for (let key of parts) {
+      // The value isn't an object which means we can't traverse it
+      if (typeof value !== 'object' || value === null) {
+        return null
+      }
+
+      // The path contains a key that doesn't exist
+      if (!(key in value)) {
```

Comment:
Would it be better to use `hasOwn` here? `hasOwn` only checks direct properties and not inherited once via the prototype chain.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn#direct_vs._inherited_properties

```suggestion
      if (!Object.hasOwn(value, key)) {
```